specs/Proposals/0040-split-states.md

```diff
@@ -0,0 +1,257 @@
+- Proposal Name: Splitting States for Simple Nightshade
+- Start Date: 2021-07-19
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+- Issue(s): [near/NEPs#225](https://github.com/near/NEPs/issues/225) [near/nearcore#4419](https://github.com/near/nearcore/issues/4419)
+
+# Summary
+[summary]: #summary
+
+This proposal proposes a way to split each shard in the blockchain into multiple shards. Currently, the near blockchain only has one shard and it needs to be split into eight shards.
+
+# Motivation
+[motivation]: #motivation
+
+To enable sharding, specifically, [Simple Nightshade](https://gov.near.org/t/road-to-simple-nightshade/1790), we need to find a way to split the current one shard state into eight shards. 
+
+
+# Guide-level explanation 
+[guide-level-explanation]: #guide-level-explanation
+
+The proposal assumes that all validators track all shards and that challenges are not enabled.
+
+Suppose the new sharding assignment come into effect at epoch T. State migration is done at epoch T-1, when the validators for epoch T are catching up states for the next epoch. At the beginning of epoch T-1, they run state sync for the current shards if needed. From the existing states, they build states for the new shards, then apply changes to the new states when they process the blocks in epoch T-1.  This whole process runs off-chain as the new states will be not included in blocks at epoch T-1. At the beginning of epoch T, the new validators start to build blocks based on the new state roots.
+
+The change involves three parts. 
+
+## Dynamic Shards
+The first issue to address in splitting shards is the assumption that the current implmentation of chain and runtime makes that the number of shards never changes. This in turn involves two parts, how the validators know when sharding changes happen and how they store different states of shards from different epochs during the transition. The former is a protocol change and the latter only affects validators' internal states.
+
+### Protocol Change
+Sharding change will be triggered as protocol version changes. Everytime a resharding is scheduled, a new protocol version number is created for the switch to new sharding. Epochs with `protocol_version` larger or equal than the threshold number will take the new sharding assignment and epochs with `protocol_version` less will take the old sharding assignment. Since the `EpochInfo` of epoch T will be determined at the end of epoch T-2, the validators will have time to prepare for states for the new shards during epoch T-1.
+
+For example, for the Simple NightShade migration, the current protocol version is 42. If there is no other pending protcol update, we can schedule the SimpleNightShadeShardSplit to happen at protocol version 43. Then when the epoch's protocol version is updated to 43, the new sharding for Simple NightShade will be enabled. 
```

Comment:
I don't think the current protocol version is 42 :) http://rpc.mainnet.near.org/status